File: cpp/algorithm/凸包问题.cpp
#include<iostream>
#include<vector>
#include<cmath>
#include<algorithm>
using namespace std;

class Point  //点类
{   
public:
    Point() {}
    Point(int x, int y):x(x), y(y) {}
    bool operator<(const Point& p)
    {
        if(x!=p.x)
            return x < p.x;
        else
            return y<p.y;  //x相同按y的大小排序
    }
    float x, y;
};       

class Convex_Hull   //凸包类
{
public:
    Convex_Hull()      //构造函数，初始化凸包
    {
        cout << "请输入点的数量：";
        cin >> n;
        cout << "输入所有点：\n";
        for(int i=0;i<n;i++)
        {
            int x, y;
            cin >> x >> y;
            points.push_back(Point(x, y));
        }
        sort(points.begin(), points.end());    //对所有点进行排序（按照横坐标从大到小）
    }

    int get_area(Point& p1, Point& p2, Point& p3)    //求三角形面积
    {
        int x1=p1.x, x2=p2.x, x3=p3.x;
        int y1=p1.y, y2=p2.y, y3=p3.y;
        return x1*y2+x3*y1+x2*y3-x3*y2-x2*y1-x1*y3;
    }

    void up_hull(int i, int j)  //对Pi和Pj连线构成的上包
    {
        int max=0, index=-1;
        for(int k=i+1;k<j;k++)
        {
            int res=get_area(points[i], points[j], points[k]);   //通过求三角形面积来找离直线距离最远的点Pmax
            if(res>max)
            {
                max = res;
                index = k;
            }
        }
        if(index!=-1)  //找到了距离直线Pi_Pj最远的点Pmax，把直线Pi_Pmax和直线Pj_Pmax作为的上面部分作为上包，进行递归
        { 
            hull_points.push_back(points[index]);   //将Pmax加入结果集
            up_hull(i, index);
            up_hull(index, j);
        }
    }
    void down_hull(int i, int j)
    {
        int max=0, index=-1;
        for(int k=j-1;k>i;k--)
        {
            int res=get_area(points[i], points[j], points[k]);   //通过求三角形面积来找离直线距离最远的点Pmax
            if(res>max)
            {
                max = res;
                index = k;
            }
        }
        if(index!=-1)  //找到了距离直线Pi_Pj最远的点Pmax，把直线Pi_Pmax和直线Pj_Pmax作为的上面部分作为上包，进行递归
        { 
            hull_points.push_back(points[index]);   //将Pmax加入结果集
            down_hull(i, index);
            down_hull(index, j);
        }
    }
    void get_hull()
    {
        up_hull(0, n-1);  //查找上凸包
        down_hull(0, n-1);  //查找下凸包
    }
    void show_result()
    {
        cout << "hello world\n";
        for(auto &it : hull_points)
        {
            cout << "(" << it.x << "," << it.y << ")"  << ", ";
        }
    }
private:
    int n;    //一共有n个点
    vector<Point> points;   //包含所有点的数组
    vector<Point> hull_points;  //凸包点
};

int main()
{
    int n;
    Convex_Hull c;
    c.get_hull();
    c.show_result();
}

File: python/凸包问题.py
import matplotlib.pyplot as plt
import math
from functools import reduce,  cmp_to_key
import operator

class Point():
    def __init__(self, x, y):
        self.x = x
        self.y = y

class Convex_hull():
    def __init__(self):
        self.points = []
        self.n = eval(input("请输入点的数量:"))
        self.visit = [0]*self.n
        self.res = []
        self.hull_points = []
        print("输入所有点:")
        for  i in range(self.n):
            x, y = input().split(" ")
            x = int(x)
            y = int(y)
            self.points.append(Point(x, y))
        self.points.sort(key=lambda p:(p.x, p.y)) #按x大小排序，x大小相同按y大小排序
        self.show_points()

    def show_res(self):
        print("凸包点：")
        for it in self.res:
            print(it)

    def show_points(self):
        for it in self.points:
            print('(', it.x, ',', it.y, ')')

    def judge(self, p1, p2, p3):                   #计算三角形面积，根据三角形面积来求顶点
        x1, x2, x3 = p1.x, p2.x, p3.x
        y1, y2, y3 = p1.y, p2.y, p3.y
        res = x1*y2 + x3*y1 + x2*y3 - x3*y2 - x2*y1 - x1*y3
        return res

    def hull(self, left, right):
        max = 0
        index = -1
        if left<right:
            for i in range(left+1, right):
                res = self.judge(self.points[left], self.points[i], self.points[right])
                if res == 0:
                    self.visit[i] = 1
                if res>max:
                    max = res
                    index = i
        else:
            for i in range(right, left):
                res = self.judge(self.points[left], self.points[i], self.points[right])
                if res == 0:
                    self.visit[i] = 1
                if res>max:
                    max = res
                    index = i
        
        if index != -1:
            self.visit[index] = 1
            self.hull(left, index)
            self.hull(index, right)
        
    def get_hull(self):
        self.visit[0] = 1
        self.visit[self.n-1] = 1
        self.hull(0, self.n-1)
        self.hull(self.n-1, 0)
        #为了将凸包点顺时针连接，这里需要作一下处理
        for i in range(self.n):
            if self.visit[i]==1:
                self.hull_points.append(Point(self.points[i].x, self.points[i].y))
                
        for it in self.hull_points:
            self.res.append((it.x, it.y))
        #将答案中的点按顺时针排序，方便画图的时候连线
        center = tuple(map(operator.truediv, reduce(lambda x, y: map(operator.add, x, y), self.res), [len(self.res)] * 2))
        self.res=sorted(self.res, key=lambda coord: (-135 - math.degrees(math.atan2(*tuple(map(operator.sub, coord, center))[::-1]))) % 360, reverse=False)

def draw_answer(points, res):
    x = []
    y = []
    x_res = []
    y_res = []
    for it in points:
        p = (it.x, it.y)
        if it not in res:
            x.append(it.x)
            y.append(it.y)
    for it in res:    
        x_res.append(it[0])
        y_res.append(it[1])
    x_res.append(res[0][0])
    y_res.append(res[0][1])
    print(x_res)
    print(y_res)
    plt.title('Convex Hull')
    plt.scatter(x, y)
    plt.scatter(x_res, y_res, c='#228B22')
    plt.plot(x_res, y_res)
    plt.show()

c = Convex_hull()
c.get_hull()     #求解凸包
c.show_res()
draw_answer(c.points, c.res)  #将结果以图形画出来
